### PR TITLE
[devel] feat: implement rolling pilot log output

### PR DESCRIPTION
**Pull request by marianne013**
@fstagni @martynia 
Simon had a go at trying to make the pilot write its log file on a rolling basis. I tested it and as far as I can tell it doesn't break anything. 
This might be a good stepping stone towards handling the output in the pilot logger.

### CODE REVIEW

**Comment by marianne013** [comment]
@sfayer Forgot to tag you.

**Comment by marianne013** [comment]
@sfayer @fstagni Simon and me were debugging some LHCb pilots (as it happens ;-) using 0 CPU yesterday and we could have really done with a rolling log output. Would you consider approving this pull request as an interim measure until the pilot logger is finished ?

**Comment by sfayer** [comment]
@marianne013: from testing, we think this patch is also needed:
```patch
             sys.stdout.write("\n")                                                                                                                                                                                                                                         
             sys.stdout.flush()                                                                                                                                                                                                                                             
+            sys.stderr.write("\n")                                                                                                                                                                                                                                         
+            sys.stderr.flush()
```

Regards,
Simon

**Comment by marianne013** [comment]
I was kind of hoping it would let me click the "implement suggestion" button, but alas, no

File: Pilot/pilotTools.py [hunk 1]
```diff
@@ -395,18 +396,37 @@ def executeAndGetOutput(self, cmd, environDict=None):
...
 
-            for line in _p.stderr:
-                sys.stderr.write(str(line))
+            # Ensure output ends on a newline
+            sys.stdout.write("\n")
+            sys.stdout.flush()
             sys.stderr.write("\n")
+            sys.stderr.flush()
+            # Decode full buffer to ascii (to avoid codepoint splitting problems)
+            outData = outData.decode("ascii", "replace")
```
[the remark this answers is not in the document]
Comment (marianne013) [reply Pilot/pilotTools.py:428]: Can you be slightly more precise as in which test is it that fails and what the error is ?

[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/pilotTools.py:428]: I extended my test to run it for a bunch of strings with different lengths (not pushed yet). It seemed to run fine, but after I repeated the test for a few times (from a command line, as shown above), I got a failure. It was somewhat difficult to see where the strings were different (at 1025 length...). A closer look (text is a random ASCII letter sequence + `\n` ) \
- the original string starts with a newline character,  the string processed by the program doesn't (it is now 1024 characters long)

So it is missing a leading newline character. Not the end of the world, we can live with it, I can exclude these cases from a test.

[the remark this answers is not in the document]
Comment (sfayer) [reply Pilot/pilotTools.py:428]: Where are we with this? We're really keen to get this rolling output in as it's hampering our attempts to debug other problems as we have to wait for the JobAgent to finish to get its log output...

Regards,
Simon

[the remark this answers is not in the document]
Comment (martynia) [reply Pilot/pilotTools.py:428]: Can you live with the fact, that the code does not work with Python 3 ? See my comments above to Daniela's post.

**Comment by martynia** [comment]
>                   outChunk = stream.read(1024).decode("ascii", "replace")
E                   AttributeError: 'str' object has no attribute 'decode'

`stream.read() `returns an `str `apparently, so the problem persists in Python 3, doesn't it ?

**Comment by sfayer** [comment]
> > ```
> >               outChunk = stream.read(1024).decode("ascii", "replace")
> > ```
> 
> E AttributeError: 'str' object has no attribute 'decode'
> 
> `stream.read() `returns an `str `apparently, so the problem persists in Python 3, doesn't it ?

Nah, stream.read() is bytes on py3, str on py2 (which is how the old code works with both)... We're still running into a few minor unicode problems, but we think that should generally fix it.

**Comment by sfayer** [comment]
We've re-tested this latest version with both py3 & py2 and it works for both.... Any other comments?

Regards,
Simon

**Review by martynia — APPROVED** [review]
Tested with Python 2 and 3. OK.

File: Pilot/pilotTools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fstagni) [reply Pilot/pilotTools.py]: > Can you live with the fact, that the code does not work with Python 3 ? See my comments above to Daniela's post.

no :)

**Comment by sfayer** [comment]
Hi @fstagni,

I think this PR is all ready to go now, do you need anything further from us to get it merged?

Regards,
Simon

**Comment by martynia** [comment]
I think we could go ahead. I’ll deal with this change in my remote logger, once it is in.

Cheers, JM

**Pull request merged**